Lib/typing.py

```diff
@@ -1060,15 +1060,24 @@ def _evaluate(self, globalns, localns, type_params=_sentinel, *, recursive_guard
                 globalns = getattr(
                     sys.modules.get(self.__forward_module__, None), '__dict__', globalns
                 )
+
+            # type parameters require some special handling,
+            # as they exist in their own scope
+            # but `eval()` does not have a dedicated parameter for that scope.
+            # For classes, names in type parameter scopes should override
+            # names in the global scope (which here are called `localns`!),
+            # but should in turn be overridden by names in the class scope
+            # (which here are called `globalns`!)
             if type_params:
-                # "Inject" type parameters into the local namespace
-                # (unless they are shadowed by assignments *in* the local namespace),
-                # as a way of emulating annotation scopes when calling `eval()`
-                locals_to_pass = {param.__name__: param for param in type_params} | localns
-            else:
-                locals_to_pass = localns
+                globalns, localns = dict(globalns), dict(localns)
+                for param in type_params:
+                    param_name = param.__name__
+                    if not self.__forward_is_class__ or param_name not in globalns:
+                        globalns[param_name] = param
+                        localns.pop(param_name, None)
```

Comment:
Why do we need to do this? Shouldn't the local namespace override the type params?

Test case:
```
class X[T]:
    T = int
    x: T
```

Reply:
I already have that test case included in this PR, and it fails unless we have this block of code here. The local namespace shoudl indeed override the type param, but the locals here are found in the `globalns`, and the globals here are found in the `localns` variable. This highly confusing situation is because the `globals` and `locals` are swapped higher up in the call stack here, for backwards compatibility reasons if you believe the comment:

https://github.com/python/cpython/blob/a3711afefa7a520b3de01be3b2367cb830d1fc84/Lib/typing.py#L2419-L2426

Reply:
I saw that comment, maybe we should find a way to get rid of that compatibility hack.

Reply:
I don't really see a way round it, short of deprecating `get_type_hints` altogether. Which... I'm not 100% against, but it would be _quite_ disruptive.

Reply:
There could be other approaches, like adding new keyword arguments. Will have to think about it more.

Reply:
If you decide to add keywords arguments or deprecate things for 3.10, please tell me in advance so that I can be reactive on Sphinx' side if needed.